### ISSUE TITLE
LXC losetup fails sometimes

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1093]
Originally reported of Google Code with ID 1037.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

# gnt-cluster --version
gnt-cluster (ganeti v2.14.0beta1) 2.14.0~beta1

# gnt-cluster version
Software version: 2.14.0~beta1
Internode protocol: 2140000
Configuration format: 2140000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version v2.14.0beta1

<b>What distribution are you using?</b>
# cat /etc/*release
SUSE Linux Enterprise Server 11 (x86_64)
VERSION = 11
PATCHLEVEL = 3

The LXC hypervisor uses CreateBdevPartitionMapping from utils/storage.py in _PrepareInstanceRootFsBdev. If the device has no partitions ReleaseBdevPartitionMapping is called to remove the losetup mapping. Sometimes the sequence of loop device mapping and deletion is to fast:

# gnt-instance start shared.local 
Waiting for job 7972 for shared.local ...
Job 7972 for shared.local has failed: Failure: command execution error:
Could not start instance 'shared.local': Hypervisor error: Failed to create partition mapping for /var/run/ganeti/instance-disks/shared.local_0: Failed to detach /dev/loop0: loop: can't delete device /dev/loop0: Device or resource busy

I can reproduce this behavior with a script, in which "losetup -d" always fails:

==8<-------------------------------
losetup -f --show /dev/mapper/some-device-without-a-partition
kpartx -a -v /dev/loop0
losetup -d /dev/loop0
==8<-------------------------------

Without further knowledge, it's always a good idea to give udev time to create/remove device files. Inserting "udevadm settle" after kpartx makes it reliable for me. 

Probably losetup can/should be skipped at all, if "image_path" is allready a block device?

Thanks, Sascha.
```
Originally added on 2015-03-09 10:12:44 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
While I cannot reproduce this using several setups, an "udevadm settle" invocation would not hurt anyone. I'll submit a patch for 2.13.

The removal of losetup invocations is a bit more tricky. If the 2.13 RC comes out soon after bug 1018 is finally resolved, this will require a fair bit more logic to allow smooth instance removal after in-place upgrades.
```
Originally added on 2015-03-10 17:00:28 +0000 UTC.

Added Labels: Priority-Medium Hypervisor-LXC SmallTask
Changed State: Accepted
Added to Milestone: Release2.13

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed in:

commit ef30bed46dcbd5e34de732996b9187c270cbc908
Author: Hrvoje Ribicic <riba@google.com>
Date:   Tue Mar 17 20:09:24 2015 +0000

    LXC: Add udevadm settle invocation to prevent errors
    
    As reported in issue 1037, LXC users could encounter a situation where
    a loopback device was busy at the time Ganeti attempted to destroy it.
    To prevent this from happening, an invocation of udevadm settle was
    added to the blockdev preparation helper function.
    
    Signed-off-by: Hrvoje Ribicic <riba@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>
```
Originally added on 2015-03-18 13:33:24 +0000 UTC.

Mentioned Issue with GoogleCode ID 1037 has been migrated to GitHub with ID #1093.
Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:03 +0000 UTC.

Changed State: Released
